File: Caper/CaperSharp/MappingCache.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.IO;

namespace CaperSharp
{
  class MappingCache
  {
    private const char Tab = '\t';

    public string ContigIdent { get; private set; }
    public int LeftIndex { get; private set; }
    public int RightIndex { get; private set; }
    public List<List<Mapping>> Sequences { get; private set; }

    public MappingCache( List<List<Mapping>> lMappings, string aContigIdent, int aLeftIndex, int aRightIndex )
    {
      Sequences = lMappings;

      ContigIdent = aContigIdent;
      LeftIndex = aLeftIndex;
      RightIndex = aRightIndex;      
    }

    

    public ICollection<Mapping> GetReads( int aLeft, int aRight )
    {
      List<Mapping> lResult = new List<Mapping>();

      for ( int i = PrivateIndex(aLeft); i <= PrivateIndex(aRight); i++ )
      {
        if ( Sequences[ i ] != null )
          lResult.AddRange( Sequences[ i ] );
      }

      return lResult;
    }

    private int PrivateIndex( int aPublicIndex )
    {
      return aPublicIndex - LeftIndex;
    }
  }
}

File: Caper/Caper/MappingCache.cpp
#include "MappingCache.h"

MappingCache::MappingCache( IndexedMappings * aIndexedReads, string & aContigIdent, long long & aLeftIndex, long long & aRightIndex )
{
  IndexedReads = aIndexedReads;

  ContigIdent = aContigIdent;
  LeftIndex = aLeftIndex;
  RightIndex = aRightIndex;
}

Mappings * MappingCache::GetReads( long long aLeft, long long aRight )
{
  Mappings * lResult = new Mappings();

  for ( long long i = PrivateIndex(aLeft); i <= PrivateIndex(aRight); ++i )
  {
    if ( !IndexedReads->at(i).empty() )
    {
      for ( long long j = 0; j < IndexedReads->at(i).size() ; ++j )
      {        
        lResult->push_back( IndexedReads->at(i)[j] );
      }
    }
  }

  return lResult;
}

int MappingCache::PrivateIndex( long long aPublicIndex )
{
  return (int) aPublicIndex - LeftIndex; // this is ok, since it'll always be < 1000. I could even do this with a short.
}

void MappingCache::DestroyMappings()
{
  for (long long i = 0; i < IndexedReads->size(); ++i )
  {
    for (long long j = 0; j < IndexedReads->at(i).size(); ++j )
    {
      delete IndexedReads->at(i)[j];
    }    
  }
}

MappingCache::~MappingCache()
{
  DestroyMappings();
  delete IndexedReads; // there's a memory leak here, because the pointers are getting deallocated...?
}


